File: python3/jump_game.py
"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)

        if n == 1:
            return True

        cred = 0
        i = 0
        while True:
            cred = max(cred, nums[i])
            if cred + i >= n-1:
                return True
            if cred == 0:
                break

            i += 1
            cred -= 1

        return False

File: golang/jump_game.go
package main

/*
Space	: O(1)
Time	: O(n)
*/

func canJump(nums []int) bool {
	n := len(nums)
	if n == 1 {
		return true
	}

	i := 1
	credit := nums[0]
	for credit > 0 {
		credit = Max(credit-1, nums[i])
		if i == n-1 {
			return true
		}
		i++
	}

	return false
}
